### PR TITLE
[TTS] EasymagpieTTS Online CFG Distillation

### PR DESCRIPTION
**Pull request by artem-gorodetskii**
## Summary

This PR adds online classifier-free guidance (CFG) distillation for EasyMagpieTTS, including local transformer (LT) distillation, phoneme guidance, rollout rejection, and single-turn/multiturn checkpoint support.

## Summary by CodeRabbit

- **New Features**
  - Added online CFG distillation stage for Easy Magpie TTS.
  - Added configurable frame-major and codebook-major ordering for distillation losses.
  - Fixed KL-divergence normalization across individual codebook/frame slices.

- **Testing**
  - Added an end-to-end functional test covering online CFG distillation training.
  - Included the new scenario in the continuous integration test suite.

### CODE REVIEW

**Comment by artem-gorodetskii** [comment]
@coderabbitai review

File: .devcontainer/devcontainer.json
Comment (XuesongYang) [inline .devcontainer/devcontainer.json:1]: we don't need to merge anything under `.devcontainer` to the main branch, right? so far as I know, it is only used for running Cursor IDE.

If your recipe requires some dependency built in docker container, we may need to add those in https://github.com/NVIDIA-NeMo/Speech/blob/main/docker/Dockerfile

Comment (artem-gorodetskii) [reply .devcontainer/devcontainer.json:1]: This is a local development environment, I will delete it when I change the PR status from draft

**Comment by artem-gorodetskii** [comment]
@coderabbitai generate sequence diagram for the training workflow inside nemo/collections/tts/models/easy_magpietts_cfg_distillation.py

File: nemo/collections/tts/models/easy_magpietts_cfg_distillation.py [hunk 1]
```diff
@@ -0,0 +1,3288 @@
...
+    dummy_embeds_uncond: Tensor
+    cache_position: Tensor
+    audio_codes: Tensor
+    audio_codes_lens: Tensor
+
+
+@dataclass
+class _GroundTruthPhonemeEmbeds:
+    embeds: Optional[Tensor]
+    lens: Optional[Tensor]
```
Comment (blisc) [inline nemo/collections/tts/models/easy_magpietts_cfg_distillation.py]: Why do you prefer to have a dataclass with two elements as opposed to returning two tensors from a function?

File: nemo/collections/tts/models/easy_magpietts_cfg_distillation.py [hunk 2]
```diff
@@ -0,0 +1,4232 @@
...
+        return self._is_active
+
+    @property
+    def use_multilingual_asr(self) -> bool:
+        """Get whether validation uses the multilingual ASR model."""
+        return self.params.use_multilingual_asr
+
+    def _setup_asr_model(self) -> None:
+        """Initialize the ASR model used for validation transcription."""
+        if self.params.use_multilingual_asr:
```
Comment (blisc) [inline nemo/collections/tts/models/easy_magpietts_cfg_distillation.py]: Why define a property if you don't plan on using it?

Either not do define the property and use params.use_multilingual_asr, or keep the property and use it.

File: nemo/collections/tts/models/easy_magpietts_cfg_distillation.py [hunk 3]
```diff
@@ -0,0 +1,4232 @@
...
+            self.whisper_model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-large-v3")
+            self.whisper_model.eval()
+            self.whisper_model.to(self.default_device)
+            self.nemo_asr_model = None
+
+            for p in self.whisper_model.parameters():
+                p.requires_grad = False
+
+        else:
+            self.nemo_asr_model = nemo_asr.models.EncDecRNNTBPEModel.from_pretrained("nvidia/parakeet-ctc-0.6b")
```
Comment (blisc) [inline nemo/collections/tts/models/easy_magpietts_cfg_distillation.py]: I think you'd be better off passing these through the params rather than hard-coding it. ctc-0.6b is also an odd choice to hard-code. I would have used 1.1b-tdt

File: nemo/collections/tts/models/easy_magpietts_cfg_distillation.py [hunk 4]
```diff
@@ -0,0 +1,4232 @@
...
+            torch.cuda.empty_cache()
+
+        self._is_active = True
+
+    def switch_device(self, device: torch.device | str) -> None:
+        """Move loaded validation models to a target device.
+
+        Args:
+            device: Target device for the loaded ASR and speaker-verification models.
+        """
```
Comment (blisc) [inline nemo/collections/tts/models/easy_magpietts_cfg_distillation.py:2451]: I'm surprised that you didn't run into my buffer issues that resulted in this function: https://github.com/NVIDIA-NeMo/Speech/pull/16035/changes#diff-313ec9e7123666a0330dead9717b70459284e851ffbf8ddb1ae886c3c72b35baR168-R173

Is it because you only load models on_fit_start() instead of init()?

File: nemo/collections/tts/models/easy_magpietts_cfg_distillation.py [hunk 5]
```diff
@@ -0,0 +1,4232 @@
...
+            embeddings=[zero_delay_tensor, embedded],
+            lengths=[delay, audio_codes_lens],
+        )
+        return embedding, lens
+
+    def _configure_training_mode(
+        self,
+        mode: str,
+        training_mode: Optional[TrainingMode],
+    ) -> TrainingMode:
```
Comment (blisc) [inline nemo/collections/tts/models/easy_magpietts_cfg_distillation.py:3112]: What does this function do?

File: nemo/collections/tts/models/easy_magpietts_cfg_distillation.py [hunk 6]
```diff
@@ -0,0 +1,4232 @@
...
+
+        Returns:
+            Mapping containing the total loss, enabled component losses and the
+            rejection-rate metric when applicable.
+        """
+        # The multiturn dataset inserts interruption markers aligned to the ground-truth audio.
+        # This distillation path is intentionally restricted to ordinary single-turn TTS cuts.
+        if self.cfg.get("use_multiturn_dataset", False):
+            text = batch["text"]
+            batch["text"] = text.masked_fill(text == self.interruption_token_id, self.pad_id)
```
Comment (blisc) [inline nemo/collections/tts/models/easy_magpietts_cfg_distillation.py:3897]: According to the comment, we do not support distillation on multi-turn. Should this code path throw an error? Does this path actually work on multi-turn data?

**Review by blisc — COMMENTED** [review]
Most of the PR looks good to me. Left a few comments, mainly on style choices. One final comment on style: Should we only keep the three main classes (_TeacherInferenceEngine, _ValidationInferenceEngine, EasyMagpieCFGDistillation‎) in the total level easy_magpietts_cfg_distillation.py and move the private dataclasses and functions elsewhere?

File: nemo/collections/tts/models/easy_magpietts_cfg_distillation.py (hunk 1 above)
Comment (artem-gorodetskii) [reply nemo/collections/tts/models/easy_magpietts_cfg_distillation.py]: I used a dataclass because the two values form one logical result and named fields made their roles explicit. It was a readability choice rather than a requirement. I’ve changed it to return a tuple as suggested.

File: nemo/collections/tts/models/easy_magpietts_cfg_distillation.py (hunk 2 above)
Comment (artem-gorodetskii) [reply nemo/collections/tts/models/easy_magpietts_cfg_distillation.py]: The property is needed because `EasyMagpieCFGDistillation` accesses validation-engine attributes and state. Internally, direct access through `params` is acceptable, the property is retained for external consumers. Kept only property.

File: nemo/collections/tts/models/easy_magpietts_cfg_distillation.py (hunk 3 above)
Comment (artem-gorodetskii) [reply nemo/collections/tts/models/easy_magpietts_cfg_distillation.py]: Agreed. I moved the ASR and speaker-verification model names into `_ValidationInferenceParams` and exposed them through the model configuration.

I kept `nvidia/parakeet-ctc-0.6b` as the default to preserve consistency with the ASR evaluator [already used](https://github.com/NVIDIA-NeMo/Speech/blob/main/nemo/collections/tts/models/easy_magpietts.py#L142) by `EasyMagpieTTSModel`. The model is no longer hard-coded: `nemo_asr_model_name` can be set to `nvidia/parakeet-tdt-1.1b` or another appropriate model in the training configuration.

File: nemo/collections/tts/models/easy_magpietts_cfg_distillation.py (hunk 4 above)
Comment (artem-gorodetskii) [reply nemo/collections/tts/models/easy_magpietts_cfg_distillation.py:2451]: Yes, but there is an additional distinction: the validation models are owned by `_ValidationInferenceEngine`, which is intentionally a plain python class rather than a `nn.Module`. Consequently, they are not registered as submodules of `EasyMagpieCFGDistillation` and DDP environment does not see or synchronize their parameters or buffers. They are also instantiated in `on_fit_start()` after DDP environment setup. Because of those two properties, I did not need to explicitly mark their parameters and buffers as DDP-ignored.

File: nemo/collections/tts/models/easy_magpietts_cfg_distillation.py (hunk 5 above)
Comment (artem-gorodetskii) [reply nemo/collections/tts/models/easy_magpietts_cfg_distillation.py:3112]: `_configure_training_mode()` resolves the `TrainingMode` used to prepare a batch.

`EasyMagpieTTSInferenceModel` supports multiple training modes and `EasyMagpieTTSModel.process_batch()` randomly selects one during training. We currently use only one mode, but I kept this behavior for compatibility and moved it into a separate helper. This also makes it easier to adopt future changes from the parent classes if we later remove the unused multi-mode functionality.

File: nemo/collections/tts/models/easy_magpietts_cfg_distillation.py (hunk 6 above)
Comment (artem-gorodetskii) [reply nemo/collections/tts/models/easy_magpietts_cfg_distillation.py:3897]: Multi-turn distillation is not supported. I added batch-level task validation that raises a `ValueError` if any sample has a task other than `tts`. We cannot reject `use_multiturn_dataset=True` itself because that representation is also used for ordinary single-turn TTS cuts. For validated TTS batches, the interruption marker is still replaced with padding, [consistent with ](https://github.com/NVIDIA-NeMo/Speech/blob/main/nemo/collections/tts/models/easy_magpietts.py#L972)`EasyMagpieTTSModel.process_batch()`.

**Comment by artem-gorodetskii** [comment]
> One final comment on style: Should we only keep the three main classes (_TeacherInferenceEngine, _ValidationInferenceEngine, EasyMagpieCFGDistillation‎) in the total level easy_magpietts_cfg_distillation.py and move the private dataclasses and functions elsewhere?

Ideally, each engine should be moved together with its private dataclasses and helper functions. For example, `_ValidationInferenceEngine`, its parameters and its metric helpers could live in a module under `nemo.collections.tts.modules.magpietts_validation`. However, I suggest handling this in a separate PR together with the broader `EasyMagpie` refactoring.

File: nemo/collections/tts/models/easy_magpietts_cfg_distillation.py (hunk 5 above)
Comment (blisc) [reply nemo/collections/tts/models/easy_magpietts_cfg_distillation.py:3112]: Ok, this is used to select how much delay we use between text, phones, and audio.

**Review by blisc — DISMISSED** [review]
We are good to merge

**Comment by blisc** [comment]
/ok to test [c5b0986](https://github.com/NVIDIA-NeMo/Speech/pull/16030/commits/c5b0986c070fdf701e85b542d74f132ada8983f9)

**Comment by blisc** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Speech/commit/73dc2e157c811e7e146c5915399359abe1bb49e9

**Review by blisc — APPROVED** [review]

**Review by chtruong814 — APPROVED** [review]
